### ISSUE TITLE
mpv doesn't select closed captions with --sid= and it makes the OSD subtitle selector behave unexpectedly

### ISSUE THREAD

**Issue by rindeal** [mpv-player/mpv#7608]
### Important Information

Provide following Information:
```sh
>>> mpv --version
mpv 0.32.0+git.20200402T120653.5824ac7d36 Copyright © 2000-2020 mpv/MPlayer/mplayer2 projects
 built on Apr  2 2020
FFmpeg library versions:
   libavutil       56.31.100
   libavcodec      58.54.100
   libavformat     58.29.100
   libswscale      5.5.100
   libavfilter     7.57.100
   libswresample   3.5.100
FFmpeg version: 4.2.2
```
openSUSE Tumbleweed

### Reproduction steps

```sh
>>> mpv -v -v --sid=4 Starship_Troopers.vob
```

### Expected behavior

mpv plays the movie with subtitle track # 4

### Actual behavior

mpv plays the movie with no subtitles and clicking the right mouse button on the subtitle indicator in OSD does nothing, <kbd>Shift+J</kbd> works fine though. Also the OSD breaks only if the selected CC track is the last one.

### Log file

Log says OSD is trying to change to track # 4, when right-clicking the indicator.

```
[cplayer] Set property: sid="4" -> 1
[cplayer] Set property: sub=4 -> 1
```

### Sample files

https://samples.ffmpeg.org/sub/subcc/ClosedCaptions/Starship_Troopers.vob

---

It's related to issue https://github.com/mpv-player/mpv/issues/3968. The selection with `--sid=` probably still won't work, because ffmpeg devs still don't expose codec properties in `AVStream`, although it's been "deprecated" for six years and their own utils still rely on `AVStream.codec` so...
But at least the OSD bug might be a new one.

**Comment by wm4** [comment]
If I understand this right, you want to set: https://mpv.io/manual/master/#options-sub-create-cc-track
Which exposes a CC track always (even if there's none), but that's the best we can do for now.

**Comment by rindeal** [comment]
> If I understand this right, you want to set: https://mpv.io/manual/master/#options-sub-create-cc-track
> Which exposes a CC track always (even if there's none), but that's the best we can do for now.

Well, `--sub-create-cc-track=yes` causes even more bugs than it solves. For the sample I linked to the behaviour of `--sid=4` changes from selecting no track and buggy OSD to selecting sub track # 3 and no buggy OSD, in either case selecting track # 4 currently seems impossible. On other samples I have, the option creates from 1 to 4 CC tracks that do not even exist and sometimes selects them as default. But to give some credits, for certain files the option really enabled CC track selection.

**Comment by wm4** [comment]
> On other samples I have, the option creates from 1 to 4 CC tracks that do not even exist 

I explicitly told you that it does that, and the manpage also says that. I'm out of here.

**Comment by rindeal** [comment]
> I'm out of here.

All right, so for the remaining folks, I've figured out why it creates multiple virtual CC tracks and only sometimes plays them by default. Several tracks are created, because mpv treats image attachments as video tracks and the option creates a virtual CC track for every one of those "video" streams. And the "sometimes default" problem is, because although the virtual tracks are marked as "default", they have lower priority than real default subtitle tracks.

I have also figured out why the OSD is unable to change sub track to the previous one. The OSD sends lua command `mp.commandv("set", "sub", 4)` / `mp.commandv("set", "sid", 4)` / `mp.set_property("sid", 4)`. This is the runtime equivalent of `--sid=4` on command line, and there is the catch. In runtime mpv checks if the new `sid` value is different from the current one and only then it changes the subtitle track. Somehow, however, mpv plays `sid=0`, but stores `sid=4` in its config. Debugging is made worse, because `mp.set_property("sid", 4)` succeeds, but `mp.get_property("sid")` returns (`-2` / `"no"`). The reason for that is, while `set_property("sid", ...)` goes through `mp_property_generic_option()`, which goes straight for config, `get_property("sid")` is intercepted in `property_switch_track()` and returns value of a runtime variable holding current track list (`mpctx->current_track`) instead of the in-memory config for `sid`.

TL;DR `m_config_t`'s `"sid"` is not properly synced with `MPContext.current_track[0][STREAM_SUB]`

**Comment by wm4** [comment]
OK, that's reasonable. I understand what this was about now.

I tried to fix some of these. Unfortunately, none of this is possible without destroying compatibility somewhat. Hopefully the impact is low.

None of this is ideal. The property/option disconnect is somewhat inherent - the idea was to remove the separation, but user-visible semantics are such that it's not possible without breaking possibly a lot of users. The main problem is really that the properties try to return the current actual selection. It's a bit too late to fix that, and maybe it was never really possible to do it without a break. What can I say, I'm sorry.

**Comment by rindeal** [comment]
Reading through your commits, I think I'm beginning to see and understand the scope of all this. One idea that occurred to me is redesigning the options in an ffmpeg-like way. That is having global and per-file options separated, both internally (i.e. single structure for global options, then a separate structure for each URL) and externally (CLI global options can be specified just once at the beginning, then per-file options before each file; config files can only specify global options and set defaults of per-file options). Also these options could be separated from core mpv functionality, just like ffmpeg libs have no idea about the options of ffmpeg CLI utils, the options could then be read-only and the core lib would rely solely on its own runtime structs. The chain would then be like mpv > libmpv > "libplacebo", with options being implemented in mpv only or split between mpv and libmpv. For reference you can look at `fftools/{cmdutils.h,cmdutils.c,ffmpeg_opt.c}`.

**Comment by wm4** [comment]
None of this helps with the case at hand. Also, ffmpeg's mechanism is terribly, both for users and in implementation, so mpv won't have this. Similar to the case at hand, runtime changes are a complication that doesn't exist in ffmpeg, because ffmpeg doesn't do option runtime changes.
